### PR TITLE
FCREPO-3834 Enable Camel toolbox to send xml records to Solr indexing service

### PR DESCRIPTION
**Pull request by Surfrdan**
** FCREPO-3834 Enable Camel toolbox to send xml records to Solr indexing service **
* * *

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3834

# What does this Pull Request do?
Removes the ldpath related code and replaces the Solr requirement on LDPath with a Solr XSLT transform implementation 

# What's new?
Replaced LDPath transform with an XSLT based transform which can be supplied per object via the same hasIndexingTransformation as the previous LDPath implementation. Alternatively, if no transofmation is specified, a bare-boines default transform is supplied on the classpathj which replicates the functaionality of the LDPath version.

# How should this be tested?

A description of what steps someone could take to:
* Make a call to the solr indexing endpoint and examing the resulting solr document whihc should be indexed on the configured solr core. 

# Additional Notes:README.md should probably be updated too to reflect the new changes and removal of the LDPath functionality, however, I wasn't sure whether that should wait until after a release as I believe the README.md is the main dicumentation for this repo and relates to the current release. 

# Interested parties
@fcrepo/committers

### CODE REVIEW

File: docker-compose/camel-toolbox-config/configuration.properties [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+jms.brokerUrl=tcp://localhost:61616
 
 solr.indexing.enabled=true
-solr.baseUrl=http://solr:8983/solr/fcrepo
+solr.baseUrl=http://localhost:8983/solr/fcrepo
+solr.fcrepo.defaultTransform=org/fcrepo/camel/indexing/solr/default_transform.xsl
 
 triplestore.indexing.enabled=true
-triplestore.baseUrl=http://fuseki:3030/fcrepo
+triplestore.baseUrl=http://localhost:3030/fcrepo
```
Comment (whikloj) [inline docker-compose/camel-toolbox-config/configuration.properties:11]: These refer to the individual containers in the `docker-compose.yml` file, by switching them to localhost it no longer works as a single `docker compose up`. I had to switch them back to test it.

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java [hunk 2]
```diff
@@ -104,6 +103,7 @@ public void configure() throws Exception {
                 .when(and(simple(config.isIndexingPredicate() + " != 'true'"),
                           simple(config.isCheckHasIndexingTransformation() + " != 'true'")))
                     .setHeader(INDEXING_TRANSFORMATION).simple(config.getDefaultTransform())
+                    .log(LoggingLevel.INFO, "sending to update_solr")
```
Comment (whikloj) [inline fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:106]: I'm thinking either remove or make DEBUG level.

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java [hunk 3]
```diff
@@ -136,48 +136,41 @@ public void configure() throws Exception {
...
+                        .log(LoggingLevel.INFO, logger,
+                            "Sending RDF for Transform with with XSLT from ${header.CamelIndexingTransformation}")
+                        .toD("xslt:${header.CamelIndexingTransformation}")
+                        .to("direct:send.to.solr")
+                    .when(header(INDEXING_TRANSFORMATION).isNotNull())
+                        .log(LoggingLevel.INFO, logger,
+                            "Sending RDF for Transform with with XSLT from ${header.CamelIndexingTransformation}")
+                        .toD("xslt:${header.CamelIndexingTransformation}")
+                        .log(LoggingLevel.INFO, logger,"${body}")
+                        .to("direct:send.to.solr")
```
Comment (whikloj) [inline fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:159]: This "when" statement and the one above it seem to perform the exact same actions except for an additional log. Could one be removed?

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java [hunk 4]
```diff
@@ -136,48 +136,41 @@ public void configure() throws Exception {
...
+                    .when(or(header(INDEXING_TRANSFORMATION).isNull(), header(INDEXING_TRANSFORMATION).isEqualTo("")))
+                        .log(LoggingLevel.INFO, logger,"No Transform supplied")
+                        .setHeader(HTTP_METHOD).constant("GET")
```
Comment (whikloj) [inline fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:162]: This `setHeader` gets changed to a POST in the `direct:send.to.solr` route. If you need it to be GET, then you might have to check for the existence of the header and only set as POST if one doesn't exist. But did you really want to GET the solr record?

**Review by whikloj — CHANGES_REQUESTED** [review]
This works just needs a few tweaks.

I do think the default LDPath gave a lot more indexing by default. Perhaps we could try to duplicate that with the default XSLT?

For example, with LDPath the resource
```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix fedora: <http://fedora.info/definitions/v4/repository#> .
@prefix ldp: <http://www.w3.org/ns/ldp#> .
@prefix dc: <http://purl.org/dc/elements/1.1/> .

<http://localhost:8080/fcrepo/rest/simple_container>
        dc:title               "A simple container" ;
        fedora:created         "2024-01-16T17:09:14.829081Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> ;
        fedora:lastModified    "2024-01-16T17:09:14.829081Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> ;
        fedora:createdBy       "fedoraAdmin" ;
        fedora:lastModifiedBy  "fedoraAdmin" ;
        rdf:type               ldp:BasicContainer ;
        rdf:type               ldp:Resource ;
        rdf:type               fedora:Resource ;
        rdf:type               ldp:RDFSource ;
        rdf:type               ldp:Container ;
        rdf:type               fedora:Container .
```
results in 
```
{
   "id":"http://fcrepo:8080/fcrepo/rest/simple_container",
   "createdBy_s":"fedoraAdmin",
   "lastModified_l":1705424954829,
   "created_l":1705424954829,
   "lastModifiedBy_s":"fedoraAdmin",
   "type_ss":["http://fedora.info/definitions/v4/repository#Container",
        "http://fedora.info/definitions/v4/repository#Resource",
        "http://www.w3.org/ns/ldp#BasicContainer",
        "http://www.w3.org/ns/ldp#Container",
        "http://www.w3.org/ns/ldp#RDFSource",
        "http://www.w3.org/ns/ldp#Resource"],
    "title":["A simple container"],
    "_version_":1788267677844766720}]
}
```

With this PR the same resource (with slightly different created/modified datetimes) results in
```
{
    "id":"http://fcrepo:8080/fcrepo/rest/simple_container",
    "_version_":1788269396095926272}]
}
```

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java (hunk 3 above)
Comment (Surfrdan) [reply fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:159]: I can't be sure what my thinking was on this but it looks like an extra debug I accidentally left in place. I'm merging these

**Review by whikloj — APPROVED** [review]
👍

**Pull request merged**